* bcgTree
Automatized phylogenetic tree building from bacterial core genomes.

[[https://travis-ci.org/iimog/bcgTree][https://travis-ci.org/iimog/bcgTree.svg?branch=master]]

An article describing bcgTree is currently under review.
If you use bcgTree for your research please cite this repository in the meantime.
Also please cite the external programs and the source of the HMMs (see Background section) if you use the default essential.hmm
** Dependencies
*** Perl
To execute bcgTree [[https://www.perl.org/][perl5]] is required along with the following modules (all available via [[http://www.cpan.org/][cpan]]):
 - Getopt::Long
 - Getopt::ArgvFile
 - Pod::Usage
 - Log::Log4perl
 - FindBin
 - File::Path
 - File::Spec
 - Bio::SeqIO
 - Bio::AlignIO
*** External programs
bcgTree is a wrapper around multiple existing tools.
The following external programs are called by bcgTree and have to be installed.
The specified versions are the ones we used for testing (older versions might or might not work).
Newer versions should work (otherwise feel free to open an issue).
 - [[http://hmmer.org/][hmmsearch]] (HMMER version 3.1b1)
 - [[http://www.drive5.com/muscle/][muscle]] (v3.8.31)
 - [[http://molevol.cmima.csic.es/castresana/Gblocks.html][Gblocks]] (version 0.91b)
 - [[http://sco.h-its.org/exelixis/web/software/raxml/][RAxML]] (version 8.2.4)
Additionally [[https://github.com/BioInf-Wuerzburg/SeqFilter][SeqFilter]] is needed but it is included as a git submodule (see Installation).
** Installation
If you have all the Perl modules and external programs installed you can get bcgTree by executing the following commands:
#+BEGIN_SRC sh
git clone --recursive https://github.com/iimog/bcgTree.git
cd bcgTree/SeqFilter
make
# Now you can run bcgTree.pl
../bin/bcgTree.pl --help
#+END_SRC
** Usage
#+BEGIN_SRC sh
Usage:
      $ bcgTree.pl [@cofigfile] --proteome bac1=bacterium1.pep.fa --proteome bac2=bacterium2.faa [options]

Options:
    [@configfile]            Optional path to a configfile with @ as prefix.
                             Config files consist of command line parameters
                             and arguments just as passed on the command
                             line. Space and comment lines are allowed (and
                             ignored). Spreading over multiple lines is
                             supported.

    --proteome <ORGANISM>=<FASTA> [--proteome <ORGANISM>=<FASTA> ..]
                             Multiple pairs of organism and proteomes as
                             fasta file paths

    [--outdir <STRING>]      output directory for the generated output files
                             (default: bcgTree)

    [--help]                 show help

    [--version]              show version number of bcgTree and exit

    [--hmmsearch-bin=<FILE>] Path to hmmsearch binary file. Default tries if
                             hmmsearch is in PATH;

    [--muscle-bin=<FILE>]    Path to muscle binary file. Default tries if
                             muscle is in PATH;

    [--gblocks-bin=<FILE>]   Path to the Gblocks binary file. Default tries
                             if Gblocks is in PATH;

    [--raxml-bin=<FILE>]     Path to the raxml binary file. Default tries if
                             raxmlHPC is in PATH;

    [--threads=<INT>]
        Number of threads to be used (currently only relevant for raxml).
        Default: 2 From the raxml man page: PTHREADS VERSION ONLY! Specify
        the number of threads you want to run. Make sure to set "-T" to at
        most the number of CPUs you have on your machine, otherwise, there
        will be a huge performance decrease!

    [--bootstraps=<INT>]
        Number of bootstraps to be used (passed to raxml). Default: 10

    [--hmmfile=<PATH>]
        Path to HMM file to be used for hmmsearch. Default:
        <bcgTreeDir>/data/essential.hmm

    [--raxml-x-rapidBootstrapRandomNumberSeed=<INT>]
        Random number seed for raxml (passed through as -x option to raxml).
        Default: Random number in range 1..1000000 (see raxml command in log
        file to find out the actual value). Note: you can abbreviate options
        (as long as they stay unique) so --raxml-x=12345 is equivalent to
        --raxml-x-rapidBootstrapRandomNumberSeed=12345

    [--raxml-p-parsimonyRandomSeed=<INT>]
        Random number seed for raxml (passed through as -p option to raxml).
        Default: Random number in range 1..1000000 (see raxml command in log
        file to find out the actual value). Note: you can abbreviate options
        (as long as they stay unique) so --raxml-p=12345 is equivalent to
        --raxml-p-parsimonyRandomSeed=12345
#+END_SRC
** Results
The results all end up in the directory specified via --outdir (or bcgTree if none is specified).
This folder contains lots of intermediate files from all steps.
If the run was successful the most interesting files will be the RAxML files:
 - <outdir>/RAxML_bestTree.final
 - <outdir>/RAxML_bipartitionsBranchLabels.final
 - <outdir>/RAxML_bipartitions.final
 - <outdir>/RAxML_bootstrap.final
 - <outdir>/RAxML_info.final
Further the log file (<outdir>/log) contains all executed commands and their output.
This is useful as a reference, for re-executing steps manually and for debugging in case something went wrong.
All other files are the outputs of different steps of the pipeline.
Their names should be self-explanatory.
** Background
107 essential genes as described in:
Dupont CL, Rusch DB, Yooseph S, et al. Genomic insights to SAR86, an abundant and uncultivated marine bacterial lineage. The ISME Journal. 2012;6(6):1186-1199. doi:10.1038/ismej.2011.189.
Supplementary Table S1 (which is actually an image) contains a list of the used genes and HMMs with cut-offs.

From the manuscript:
"Genome completeness estimates
Using the Comprehensive Microbial Resource as a database, 107 hidden Markov models (HMMs) that hit
only one gene in greater than 95% of bacterial genomes were identified (Supplementary Table S1).
Trusted cutoff scores for the TIGRFAMs and Pfam HMMs were those supplied by the 
TIGRFAMs and Pfam libraries (Haft et al., 2003; Finn et al., 2010)."

In the publication:
M Albertsen,	Hugenholtz P, Skarshewski A, Nielsen KL, Tyson GW and Nielsen PH, Genome sequences of rare, uncultured bacteria obtained by differential coverage binning of multiple metagenomes. Nature Biotechnology 31, 533â€“538 (2013) doi:10.1038/nbt.2579
the authors use the same list of 107 genes (111 HMMs, glyS, pheT, proS and rpoC have two HMMs each)
as above and provide a readily created hmm file via [[https://github.com/MadsAlbertsen/multi-metagenome/][GitHub]].
This file has been used as a starting point but an [[https://github.com/MadsAlbertsen/multi-metagenome/issues/15][error]] had to be fixed.
